src/Databases/DatabasesCommon.cpp

```diff
@@ -125,16 +126,43 @@ void DatabaseWithOwnTablesBase::shutdown()
         tables_snapshot = tables;
     }
 
-    for (const auto & kv : tables_snapshot)
+    auto shutdown_table = [&](const StoragePtr & table)
     {
-        auto table_id = kv.second->getStorageID();
-        kv.second->shutdown();
+        auto table_id = table->getStorageID();
+        table->shutdown();
         if (table_id.hasUUID())
         {
             assert(getDatabaseName() == DatabaseCatalog::TEMPORARY_DATABASE || getUUID() != UUIDHelpers::Nil);
             DatabaseCatalog::instance().removeUUIDMapping(table_id.uuid);
         }
+    };
+
+    /// Destroy Buffer tables first, to flush the data to the underlying storage.
+    std::vector<StoragePtr> buffer_storages;
+    std::vector<StoragePtr> other_storages;
+
+    for (const auto & [_, storage] : tables_snapshot)
+    {
+        if (dynamic_cast<StorageBuffer *>(storage.get()))
```

Comment:
@azat  We do not use dynamic_cast in new code.

Reply:
Consider to add IStorage interface method bool isBuffer().

Reply:
News to me... How is `isBuffer()` better?

Reply:
@akuzm.
1. Easier to find in project.
2. I see that there are such interface method, so I can predict that some client will use this.
3. I can put StorageBuffer in separate .cpp file, with dynamic cast I need class definition.

Reply:
I mean, we're not philosophers building an ontology of things, so we don't really care what it "is". What this method actually says is "enable some hacks, I don't know which ones, I don't know where". This doesn't really explain much, and only makes the interface less clear. Maybe it's better because we can at least grep for it...

Reply:
@akuzm, we actually plan to move all storages, dictionaries, functions in separate .cpp files dynamic_cast does not allow to do that.
There are already methods isView, isDictionary, isBuffer method will not make interface less clear. Actually it seems that our IStorage is sort of relation in postgres. Correct me if I am wrong.

Reply:
Yeah, I'm not really opposed to it, just suddenly realized that all identity statements are a lie and these methods actually mean `enableUnknownHacksForDictionariesSomewhere`.

Reply:
By the way, do we ultimately want to have explicit dependencies between storages? I vaguely recall there was also a problem with initalization of dictionaries vs tables they look at...

Reply:
@akuzm related to explicit dependencies, you can check I removed all hacks and made dictionary look like IStorage.
If dictionary point to table from same database there will be no problem. If dictionary points to table from other database, there also will be no problem because during startup dictionaries are initialized lazilly, if client specify option lazy_dictionaries = false, than during startup dictionary load will fail, but it will not break server startup, so later it will be loaded.
Only problem is if table has dictGet() in definition, for example in column definition, to get type of dictGet we need to execute it, so it can break server startup, the only workaround right now is to specify type manually dictGetUInt64.
There is other issues. @azat are there are issues with dependencies with Buffer tables ?
There is ticket for topological sort that can solve such problems, but probably it is not first priority.

Reply:
>are there are issues with dependencies with Buffer tables ?

Little one (that I know about), but there is a patch already - #24068

>There is ticket for topological sort that can solve such problems, but probably it is not first priority.

Yeah, I also though about more generic solution (like dependency manager), but it is complex for now.